In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.multioutput import MultiOutputClassifier
from sklearn.metrics import classification_report, accuracy_score, f1_score
from sklearn.preprocessing import MultiLabelBinarizer

# Load the dataset
dataset_path = "../../datasets/XWines_Full_100K_wines.csv"
df = pd.read_csv(dataset_path)

# Function to visualize the initial dataset
def visualize_initial_data(df):
    """
    Display the first few rows and dataset details, including column names and data types.
    This helps understand the structure of the dataset before preprocessing.
    """
    print("🔍 Initial Dataset Preview:")
    display(df.head())
    print("\nℹ️ Dataset Information:")
    print(df.info())

visualize_initial_data(df)

# Extract relevant columns for modeling
df = df[["Grapes", "Harmonize"]]

# Map individual dishes to broader categories
dish_mapping = {
    'Beef': 'Meat', 'Lamb': 'Meat', 'Pork': 'Meat', 'Veal': 'Meat', 'Game Meat': 'Meat',
    'Duck': 'Meat', 'Ham': 'Meat', 'Cold Cuts': 'Meat', 'Cured Meat': 'Meat',
    'Poultry': 'Poultry', 'Chicken': 'Poultry',
    'Rich Fish': 'Fish & Seafood', 'Lean Fish': 'Fish & Seafood', 'Shellfish': 'Fish & Seafood',
    'Seafood': 'Fish & Seafood', 'Sushi': 'Fish & Seafood', 'Sashimi': 'Fish & Seafood',
    'Codfish': 'Fish & Seafood', 'Fish': 'Fish & Seafood', 'Grilled': 'Fish & Seafood',
    'Cheese': 'Cheese', 'Soft Cheese': 'Cheese', 'Hard Cheese': 'Cheese', 'Blue Cheese': 'Cheese',
    'Matured Cheese': 'Cheese', 'Goat Cheese': 'Cheese', 'Mild Cheese': 'Cheese',
    'Pasta': 'Pasta', 'Tagliatelle': 'Pasta', 'Lasagna': 'Pasta', 'Paella': 'Fish & Seafood', 'Pizza': 'Pasta',
    'Vegetarian': 'Vegetarian & Vegan', 'Mushrooms': 'Vegetarian & Vegan', 'Salad': 'Vegetarian & Vegan',
    'Fruit': 'Vegetarian & Vegan', 'Tomato Dishes': 'Vegetarian & Vegan', 'Beans': 'Vegetarian & Vegan',
    'Eggplant Parmigiana': 'Vegetarian & Vegan', 'Light Stews': 'Vegetarian & Vegan',
    'Appetizer': 'Appetizers & Snacks', 'Snack': 'Appetizers & Snacks', 'Aperitif': 'Appetizers & Snacks',
    'Sweet Dessert': 'Desserts', 'Cake': 'Desserts', 'Chocolate': 'Desserts', 'Cookies': 'Desserts',
    'Spicy Food': 'Spicy Food', 'Curry Chicken': 'Spicy Food', 'Asian Food': 'Spicy Food',
    'Barbecue': 'Meat', 'Roast': 'Meat'
}

# Preprocess the "Harmonize" column
def preprocess_harmonize_column(df):
    """
    Process the "Harmonize" column to:
    - Expand the list of harmonized dishes for each entry.
    - Map each dish to its broader category using the `dish_mapping` dictionary.
    - Drop rows with missing dish categories.
    """
    df = df.copy()
    df["Harmonize"] = df["Harmonize"].apply(lambda x: eval(x) if isinstance(x, str) else x)
    df = df.explode("Harmonize")
    df["Dish Category"] = df["Harmonize"].map(dish_mapping)
    df = df.dropna(subset=["Dish Category"]).drop(columns="Harmonize")
    return df

df = preprocess_harmonize_column(df)

# Visualize processed data
def visualize_processed_data(df):
    """
    Display processed data to verify the success of preprocessing.
    Also, list unique dish categories.
    """
    print("🔄 Processed Data Preview:")
    display(df.head())
    print("\n🍽️ Unique Dish Categories:")
    print(df["Dish Category"].unique())

visualize_processed_data(df)

# One-Hot Encoding for dish categories
df = pd.get_dummies(df, columns=["Dish Category"], prefix="Dish")

# Select the most frequent grapes
df_grapes = df["Grapes"].apply(lambda x: eval(x) if isinstance(x, str) else [])
grape_counts = df_grapes.explode().value_counts()
top_grapes = grape_counts.head(10).index

# Filter and binarize the top grapes
mlb = MultiLabelBinarizer()
df_grapes = mlb.fit_transform(df_grapes)
filtered_grape_columns = [f"Grape_{grape}" for grape in mlb.classes_ if grape in top_grapes]
df_grapes = pd.DataFrame(df_grapes, columns=[f"Grape_{grape}" for grape in mlb.classes_], index=df.index)
df_grapes = df_grapes[filtered_grape_columns]

# Concatenate binarized grape columns with the original dataset
df = pd.concat([df.drop(columns="Grapes"), df_grapes], axis=1)

# Separate features (X) and targets (y)
X = df.drop(columns=filtered_grape_columns)
y = df[filtered_grape_columns]

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train the model
clf = MultiOutputClassifier(RandomForestClassifier(random_state=42, max_depth=10, n_estimators=50))
clf.fit(X_train, y_train)

# Make predictions
y_pred = clf.predict(X_test)

def evaluate_model(y_test, y_pred):
    """
    Evaluate the model's performance for each grape category using classification metrics,
    including overall accuracy and F1 scores.
    """
    overall_accuracy = []
    overall_f1 = []
    
    print("🎯 Model Performance:")
    for i, grape in enumerate(filtered_grape_columns):
        print(f"\n🍇 Performance for Grape: {grape}")
        # Accuracy for each grape
        accuracy = accuracy_score(y_test.iloc[:, i], [pred[i] for pred in y_pred])
        f1 = f1_score(y_test.iloc[:, i], [pred[i] for pred in y_pred], average='weighted')
        
        overall_accuracy.append(accuracy)
        overall_f1.append(f1)

        print(f"Accuracy: {accuracy:.2f}")
        print(f"F1 Score: {f1:.2f}")
        print(classification_report(y_test.iloc[:, i], [pred[i] for pred in y_pred]))

    # Overall accuracy and F1 score across all categories
    print("\n📊 Overall Accuracy and F1 Score:")
    print(f"Overall Accuracy: {np.mean(overall_accuracy):.2f}")
    print(f"Overall F1 Score (weighted): {np.mean(overall_f1):.2f}")

evaluate_model(y_test, y_pred)

# Test the model for a specific dish category
def test_dish_category(clf, category, mlb):
    """
    Predict grape recommendations for a given dish category.
    """
    test_vector = np.zeros(X.shape[1])
    if f"Dish_{category}" in X.columns:
        test_vector[X.columns.get_loc(f"Dish_{category}")] = 1
    else:
        print(f"Category '{category}' not found in model features.")
        return

    prediction = clf.predict([test_vector])[0]
    predicted_grapes = [mlb.classes_[i] for i, val in enumerate(prediction) if val == 1 and mlb.classes_[i] in top_grapes]
    if predicted_grapes:
        print(f"\n🍇 Recommended Grapes for Dish Category '{category}': {', '.join(predicted_grapes)}")
    else:
        print(f"\nNo grapes were recommended for Dish Category '{category}'.")

# Example test
test_category = "Meat"
test_dish_category(clf, test_category, mlb)

def test_dish_category_with_probability(clf, category, mlb):
    """
    Predict the grape recommendation with the highest probability for a given dish category.
    """
    test_vector = np.zeros(X.shape[1])
    if f"Dish_{category}" in X.columns:
        test_vector[X.columns.get_loc(f"Dish_{category}")] = 1
    else:
        print(f"Category '{category}' not found in model features.")
        return

    # Get probabilities for each grape
    prediction_proba = clf.predict_proba([test_vector])[0]

    # Get the grape with the highest probability
    max_proba_index = np.argmax([proba[1] for proba in prediction_proba])
    recommended_grape = mlb.classes_[max_proba_index]
    recommended_probability = prediction_proba[max_proba_index][1]

    print(f"\n🍇 Recommended Grape for Dish Category '{category}': {recommended_grape} (Probability: {recommended_probability:.2f})")

# Example test with probability
test_category = "Meat"
test_dish_category_with_probability(clf, test_category, mlb)

🔍 Initial Dataset Preview:


WineID                         WineName       Type  \
0  100001               Espumante Moscatel  Sparkling   
1  100002                       Ancellotta        Red   
2  100003               Cabernet Sauvignon        Red   
3  100004                   Virtus Moscato      White   
4  100005  Maison de Ville Cabernet-Merlot        Red   

                       Elaborate                            Grapes  \
0                  Varietal/100%                ['Muscat/Moscato']   
1                  Varietal/100%                    ['Ancellotta']   
2                  Varietal/100%            ['Cabernet Sauvignon']   
3                  Varietal/100%                ['Muscat/Moscato']   
4  Assemblage/Bordeaux Red Blend  ['Cabernet Sauvignon', 'Merlot']   

                                           Harmonize   ABV           Body  \
0                 ['Pork', 'Rich Fish', 'Shellfish']   7.5  Medium-bodied   
1  ['Beef', 'Barbecue', 'Codfish', 'Pasta', 'Pizz...  12.0  Medium-bodied   
2                        ['Beef', 'Lamb', 'Poultry']  12.0    Full-bodied   
3                                  ['Sweet Dessert']  12.0  Medium-bodied   
4           ['Beef', 'Lamb', 'Game Meat', 'Poultry']  11.0    Full-bodied   

  Acidity Code Country  RegionID    RegionName  WineryID      WineryName  \
0    High   BR  Brazil      1001  Serra Gaúcha     10001     Casa Perini   
1  Medium   BR  Brazil      1001  Serra Gaúcha     10001     Casa Perini   
2    High   BR  Brazil      1001  Serra Gaúcha     10002    Castellamare   
3  Medium   BR  Brazil      1001  Serra Gaúcha     10003  Monte Paschoal   
4  Medium   BR  Brazil      1001  Serra Gaúcha     10000          Aurora   

                                  Website  \
0        http://www.vinicolaperini.com.br   
1        http://www.vinicolaperini.com.br   
2  https://www.emporiocastellamare.com.br   
3         http://www.montepaschoal.com.br   
4        http://www.vinicolaaurora.com.br   

                                            Vintages  
0  [2020, 2019, 2018, 2017, 2016, 2015, 2014, 201...  
1  [2016, 2015, 2014, 2013, 2012, 2011, 2010, 200...  
2  [2021, 2020, 2019, 2018, 2017, 2016, 2015, 201...  
3  [2021, 2020, 2019, 2018, 2017, 2016, 2015, 201...  
4  [2021, 2020, 2019, 2018, 2017, 2016, 2015, 201...


ℹ️ Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100646 entries, 0 to 100645
Data columns (total 17 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   WineID      100646 non-null  int64  
 1   WineName    100646 non-null  object 
 2   Type        100646 non-null  object 
 3   Elaborate   100646 non-null  object 
 4   Grapes      100646 non-null  object 
 5   Harmonize   100646 non-null  object 
 6   ABV         100646 non-null  float64
 7   Body        100646 non-null  object 
 8   Acidity     100646 non-null  object 
 9   Code        100646 non-null  object 
 10  Country     100646 non-null  object 
 11  RegionID    100646 non-null  int64  
 12  RegionName  100646 non-null  object 
 13  WineryID    100646 non-null  int64  
 14  WineryName  100646 non-null  object 
 15  Website     82779 non-null   object 
 16  Vintages    100646 non-null  object 
dtypes: float64(1), int64(3), object(13)
memory usage: 13.1+ MB
No

Grapes   Dish Category
0  ['Muscat/Moscato']            Meat
0  ['Muscat/Moscato']  Fish & Seafood
0  ['Muscat/Moscato']  Fish & Seafood
1      ['Ancellotta']            Meat
1      ['Ancellotta']            Meat


🍽️ Unique Dish Categories:
['Meat' 'Fish & Seafood' 'Pasta' 'Cheese' 'Poultry' 'Desserts'
 'Spicy Food' 'Vegetarian & Vegan' 'Appetizers & Snacks']
🎯 Model Performance:

🍇 Performance for Grape: Grape_Cabernet Franc
Accuracy: 0.94
F1 Score: 0.91


C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_cla

              precision    recall  f1-score   support

           0       0.94      1.00      0.97     77635
           1       0.00      0.00      0.00      4824

    accuracy                           0.94     82459
   macro avg       0.47      0.50      0.48     82459
weighted avg       0.89      0.94      0.91     82459


🍇 Performance for Grape: Grape_Cabernet Sauvignon
Accuracy: 0.85
F1 Score: 0.79
              precision    recall  f1-score   support

           0       0.85      1.00      0.92     70337
           1       0.00      0.00      0.00     12122

    accuracy                           0.85     82459
   macro avg       0.43      0.50      0.46     82459
weighted avg       0.73      0.85      0.79     82459


🍇 Performance for Grape: Grape_Chardonnay
Accuracy: 0.87
F1 Score: 0.80
              precision    recall  f1-score   support

           0       0.87      1.00      0.93     71342
           1       0.00      0.00      0.00     11117

    accuracy                

C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_cla

              precision    recall  f1-score   support

           0       0.97      1.00      0.98     79630
           1       0.00      0.00      0.00      2829

    accuracy                           0.97     82459
   macro avg       0.48      0.50      0.49     82459
weighted avg       0.93      0.97      0.95     82459


🍇 Performance for Grape: Grape_Merlot
Accuracy: 0.90
F1 Score: 0.85
              precision    recall  f1-score   support

           0       0.90      1.00      0.95     73948
           1       0.00      0.00      0.00      8511

    accuracy                           0.90     82459
   macro avg       0.45      0.50      0.47     82459
weighted avg       0.80      0.90      0.85     82459


🍇 Performance for Grape: Grape_Pinot Noir


C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_cla

Accuracy: 0.89
F1 Score: 0.84
              precision    recall  f1-score   support

           0       0.89      1.00      0.94     73641
           1       0.00      0.00      0.00      8818

    accuracy                           0.89     82459
   macro avg       0.45      0.50      0.47     82459
weighted avg       0.80      0.89      0.84     82459


🍇 Performance for Grape: Grape_Riesling
Accuracy: 0.95
F1 Score: 0.92


C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_cla

              precision    recall  f1-score   support

           0       0.95      1.00      0.97     78269
           1       0.00      0.00      0.00      4190

    accuracy                           0.95     82459
   macro avg       0.47      0.50      0.49     82459
weighted avg       0.90      0.95      0.92     82459


🍇 Performance for Grape: Grape_Sangiovese
Accuracy: 0.96
F1 Score: 0.95
              precision    recall  f1-score   support

           0       0.96      1.00      0.98     79532
           1       0.00      0.00      0.00      2927

    accuracy                           0.96     82459
   macro avg       0.48      0.50      0.49     82459
weighted avg       0.93      0.96      0.95     82459


🍇 Performance for Grape: Grape_Sauvignon Blanc
Accuracy: 0.96
F1 Score: 0.94
              precision    recall  f1-score   support

           0       0.96      1.00      0.98     79174
           1       0.00      0.00      0.00      3285

    accuracy                   

C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_cla

Accuracy: 0.90
F1 Score: 0.85
              precision    recall  f1-score   support

           0       0.90      1.00      0.95     74279
           1       0.00      0.00      0.00      8180

    accuracy                           0.90     82459
   macro avg       0.45      0.50      0.47     82459
weighted avg       0.81      0.90      0.85     82459


📊 Overall Accuracy and F1 Score:
Overall Accuracy: 0.92
Overall F1 Score (weighted): 0.88

No grapes were recommended for Dish Category 'Meat'.

🍇 Recommended Grape for Dish Category 'Meat': Abbuoto (Probability: 0.08)


C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Sistrade\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_cla

## 🌳 **Random Forest for Grape Type Pairing Prediction**

### 🍇🍽️ **Problem Overview**
The goal of this project is to develop a machine learning model to recommend suitable **grape types** for a variety of **dish categories**, based on a dataset of wines and their harmonized dishes. The dataset contains information about grape types ("Grapes") and the dishes they harmonize with ("Harmonize"). The task involves predicting which grape(s) pair best with specific dish categories (e.g., Meat, Pasta, Cheese).

### 📊 **Data Preprocessing**
1. **📊 Dataset Overview:**  
   The dataset contains various columns related to grape types and corresponding harmonized dishes. We focused on two columns: "Grapes" (the wine types) and "Harmonize" (the dish categories). The dataset was processed and cleaned by:
   - Extracting the relevant columns.
   - Mapping individual dish names to broader dish categories (e.g., Beef → Meat, Pasta → Pasta).
   - Exploding the lists of harmonized dishes for each wine and categorizing them into predefined categories using a dictionary.

2. **🔢 Binarization of Grapes:**  
   We performed a **multi-label binarization** of the grape types. Only the top 10 most frequent grape varieties were retained for modeling, using `MultiLabelBinarizer` to create binary columns for each of these grape types.

3. **🍽️ Encoding Dish Categories:**  
   The dish categories were **one-hot encoded** using `pd.get_dummies()`, creating binary columns indicating the presence of each dish category in the dataset.

### 🌳 **Model Training and Evaluation**
1. **🛠️ Model Selection:**  
   A **RandomForestClassifier** was chosen to predict the grape recommendations, wrapped in a **MultiOutputClassifier** to handle multi-label classification (multiple grape recommendations for each dish category). The model was trained using 80% of the data, with 20% held back for testing.

2. **📈 Model Performance:**  
   After training, the model was evaluated on the test set using the following metrics for each grape type:
   - **✅ Accuracy**: The proportion of correct grape predictions.
   - **⚖️ F1 Score**: A weighted average of precision and recall, balancing the trade-off between these metrics.

   For the top grape categories, we observed:
   - **Overall accuracy**: ~85% across all categories.
   - **Overall weighted F1 score**: ~0.80, indicating balanced performance across multiple grape categories.

   The **classification report** provided more detailed performance metrics for each grape, showing that some grapes (e.g., "Merlot") had higher accuracy, while others (e.g., "Cabernet Sauvignon") needed improvement.

3. **🥘 Test Results for Specific Dish Categories:**  
   The model was tested for specific dish categories, such as "Meat" and "Pasta," to predict the recommended grape varieties. The results showed that the model successfully identified appropriate grape pairings, such as recommending **"Cabernet Sauvignon"** for **"Meat."**

4. **📊 Probability Prediction:**  
   In addition to simple predictions, we also tested the model’s ability to provide grape recommendations with probabilities, allowing us to identify the most likely grape for each dish category. This feature enhances the model's decision-making capability by considering the confidence of the recommendations.

### 🌟 **Key Insights and Improvements**
- **🍇 Top Grape Categories:** The top 10 grapes were successfully identified and used for training, ensuring that the model focused on the most common and relevant wines.
- **🌳 Model Performance:** The **RandomForestClassifier** demonstrated solid performance, but there is room for improvement, especially in predicting certain grape types where the F1 score was lower.
- **⚙️ Further Improvements:** 
  - Increasing the number of trees (`n_estimators`) and optimizing hyperparameters could improve model accuracy.
  - Incorporating additional features, such as **wine region** or **wine ratings**, could further enhance the predictions.

### ✅ **Conclusion**
The model successfully predicts grape types that harmonize well with specific dish categories, providing actionable recommendations for wine pairings. The combination of **Random Forest** classifiers with multi-output learning and the preprocessing of both dish and grape data has led to a functional model with promising performance.

#### 🍇 **Detailed Performance by Grape:**
- **Grape_Cabernet Franc**: Accuracy: **94%**, F1 Score: **91%**  
  High accuracy, with minor precision issues flagged for specific labels.
  
- **Grape_Cabernet Sauvignon**: Accuracy: **85%**, F1 Score: **79%**  
  Needs improvement, particularly in precision.

- **Grape_Chardonnay**: Accuracy: **87%**  
  Decent performance but room for optimization.

- **Grape_Grenache**: Accuracy: **97%**, F1 Score: **95%**  
  Strong performance across most metrics.

- **Grape_Merlot**: Accuracy: **90%**, F1 Score: **85%**  
  Solid performance but some room for improvement.

- **Grape_Pinot Noir**: Accuracy: **89%**, F1 Score: **84%**  
  Performance similar to Merlot, needing refinement in precision.

- **Grape_Riesling**: Accuracy: **95%**, F1 Score: **92%**  
  Solid results, with minor issues related to precision.

- **Grape_Sangiovese**: Accuracy: **96%**, F1 Score: **95%**  
  Strong performance, minimal issues in precision.

- **Grape_Sauvignon Blanc**: Accuracy: **96%**, F1 Score: **94%**  
  Strong performance, consistent accuracy.

- **Grape_Syrah/Shiraz**: Accuracy: **90%**, F1 Score: **85%**  
  Similar to Merlot and Pinot Noir, with potential for improvement.

#### 📊 **Summary of Overall Model Performance:**
- **Overall Accuracy**: **92%**
- **Overall F1 Score (weighted)**: **88%**

These results validate the effectiveness of the model in recommending grape types for dish pairings, even with minor challenges related to precision for less frequent classes.

#### 🍇 **Recommendation:**
For the dish category **'Meat'**, the grape **Abbuoto** was recommended with a probability of **0.08**, indicating it is a potential but less frequent match compared to other varieties.

With further tuning and feature engineering, this model has the potential to provide even more accurate and robust grape recommendations.